### ISSUE TITLE
StopIteration in get_rope_index for Qwen3-VL video training with transformers==5.3.0

### ISSUE THREAD

**Issue by toannguyen1904** [2U1/Qwen2-VL-Finetune#250]
**Error:**

```
File "transformers/models/qwen3_vl/modeling_qwen3_vl.py", line 1126, in get_rope_index
    grid_thw = next(grid_iters[modality_type])
StopIteration
```
**Root cause:**

The Qwen3-VL processor in `transformers==5.3.0` generates `mm_token_type_ids` with per-frame type-2 groups for video tokens. Each temporal frame is a separate contiguous block of type-2 (video) tokens, separated by type-0 (text) timestamp tokens:

```
[type-2 × 144] [type-0 × 8] [type-2 × 144] [type-0 × 8] ...
   frame 1       timestamp      frame 2       timestamp
```
However, `video_grid_thw` has one entry per video, not per frame. For a video with grid `(T=20, H=18, W=32)`, the processor produces 20 contiguous type-2 groups but only 1 grid entry.

`get_rope_index` (line 1096-1126) creates a shared iterator over `video_grid_thw` and calls `next()` for each contiguous type-2 group it encounters. The iterator is exhausted after the first frame, and `StopIteration` is raised on frame 2.

**Suggested fix made by Opus4.6:**

In `monkey_patch_forward.py`, expand `video_grid_thw` from per-video `(T, H, W)` to per-frame `(1, H, W)` entries only for the `compute_3d_position_ids` call. The original per-video grid is preserved for `get_video_features` (visual encoder needs the full temporal structure for 3D convolution):

```
def _expand_video_grid_to_frames(video_grid_thw):
    if video_grid_thw is None:
        return None
    frames = []
    for grid in video_grid_thw:
        t = grid[0].item()
        frame_grid = grid.unsqueeze(0).expand(t, -1).clone()
        frame_grid[:, 0] = 1
        frames.append(frame_grid)
    return torch.cat(frames, dim=0)
```

Applied in `qwen3_vl_mixed_modality_forward` and `qwen3_vl_moe_mixed_modality_forward`:

```
# visual encoder uses original per-video grid
video_outputs = self.get_video_features(pixel_values_videos, video_grid_thw, ...)
# position computation uses expanded per-frame grid
position_ids = self.compute_3d_position_ids(
    ...,
    video_grid_thw=_expand_video_grid_to_frames(video_grid_thw),
    ...
)
```

**Verified:** Training runs successfully after the fix (forward + backward pass, loss decreasing).

**Comment by 2U1** [comment]
Thanks! I'll look into it and fix as you suggested.

**Comment by 2U1** [comment]
Thanks the qwen3.5 has the same problem, so added this into qwen3.5 too.
